File: src/main/java/me/zeroeightsix/kami/module/modules/movement/Jesus.java
package me.zeroeightsix.kami.module.modules.movement;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.KamiEvent;
import me.zeroeightsix.kami.event.events.AddCollisionBoxToListEvent;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.module.modules.player.Freecam;
import me.zeroeightsix.kami.util.EntityUtil;
import me.zeroeightsix.kami.util.Wrapper;
import net.minecraft.block.BlockLiquid;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityBoat;
import net.minecraft.network.play.client.CPacketPlayer;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.MathHelper;

import static me.zeroeightsix.kami.KamiMod.MODULE_MANAGER;

/**
 * Created by 086 on 11/12/2017.
 */
@Module.Info(name = "Jesus", description = "Allows you to walk on water", category = Module.Category.MOVEMENT)
public class Jesus extends Module {

    private static final AxisAlignedBB WATER_WALK_AA = new AxisAlignedBB(0.D, 0.D, 0.D, 1.D, 0.99D, 1.D);

    @Override
    public void onUpdate() {
        if (!MODULE_MANAGER.isModuleEnabled(Freecam.class)) {
            if (EntityUtil.isInWater(mc.player) && !mc.player.isSneaking()) {
                mc.player.motionY = 0.1;
                if (mc.player.getRidingEntity() != null && !(mc.player.getRidingEntity() instanceof EntityBoat)) {
                    mc.player.getRidingEntity().motionY = 0.3;
                }
            }
        }
    }

    @EventHandler
    Listener<AddCollisionBoxToListEvent> addCollisionBoxToListEventListener = new Listener<>((event) -> {
        if (mc.player != null
                && (event.getBlock() instanceof BlockLiquid)
                && (EntityUtil.isDrivenByPlayer(event.getEntity()) || event.getEntity() == mc.player)
                && !(event.getEntity() instanceof EntityBoat)
                && !mc.player.isSneaking()
                && mc.player.fallDistance < 3
                && !EntityUtil.isInWater(mc.player)
                && (EntityUtil.isAboveWater(mc.player, false) || EntityUtil.isAboveWater(mc.player.getRidingEntity(), false))
                && isAboveBlock(mc.player, event.getPos())) {
            AxisAlignedBB axisalignedbb = WATER_WALK_AA.offset(event.getPos());
            if (event.getEntityBox().intersects(axisalignedbb)) event.getCollidingBoxes().add(axisalignedbb);
            event.cancel();
        }
    });

    @EventHandler
    Listener<PacketEvent.Send> packetEventSendListener = new Listener<>(event -> {
        if (event.getEra() == KamiEvent.Era.PRE) {
            if (event.getPacket() instanceof CPacketPlayer) {
                if (EntityUtil.isAboveWater(mc.player, true) && !EntityUtil.isInWater(mc.player) && !isAboveLand(mc.player)) {
                    int ticks = mc.player.ticksExisted % 2;
                    if (ticks == 0) ((CPacketPlayer) event.getPacket()).y += 0.02D;
                }
            }
        }
    });

    @SuppressWarnings("deprecation")
    private static boolean isAboveLand(Entity entity) {
        if (entity == null) return false;

        double y = entity.posY - 0.01;

        for (int x = MathHelper.floor(entity.posX); x < MathHelper.ceil(entity.posX); x++)
            for (int z = MathHelper.floor(entity.posZ); z < MathHelper.ceil(entity.posZ); z++) {
                BlockPos pos = new BlockPos(x, MathHelper.floor(y), z);

                if (Wrapper.getWorld().getBlockState(pos).getBlock().isFullBlock(Wrapper.getWorld().getBlockState(pos)))
                    return true;
            }

        return false;
    }

    private static boolean isAboveBlock(Entity entity, BlockPos pos) {
        return entity.posY >= pos.getY();
    }

}

File: src/main/java/me/zeroeightsix/kami/module/modules/movement/Jesus.kt
package me.zeroeightsix.kami.module.modules.movement

import me.zeroeightsix.kami.event.events.AddCollisionBoxToListEvent
import me.zeroeightsix.kami.event.events.PacketEvent
import me.zeroeightsix.kami.mixin.extension.y
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.ModuleConfig.setting
import me.zeroeightsix.kami.util.BaritoneUtils
import me.zeroeightsix.kami.util.EntityUtils
import me.zeroeightsix.kami.util.threads.safeListener
import net.minecraft.block.BlockLiquid
import net.minecraft.entity.Entity
import net.minecraft.entity.item.EntityBoat
import net.minecraft.network.play.client.CPacketPlayer
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.gameevent.TickEvent
import org.kamiblue.commons.extension.ceilToInt
import org.kamiblue.commons.extension.floorToInt
import org.kamiblue.event.listener.listener

@Module.Info(
    name = "Jesus",
    description = "Allows you to walk on water",
    category = Module.Category.MOVEMENT
)
object Jesus : Module() {

    private val dolphin = setting("Dolphin", false)

    private val WATER_WALK_AA = AxisAlignedBB(0.0, 0.0, 0.0, 1.0, 0.99, 1.0)

    override fun onToggle() {
        BaritoneUtils.settings?.assumeWalkOnWater?.value = isEnabled
    }

    init {
        safeListener<TickEvent.ClientTickEvent> {
            if (isInWater(player) && !player.isSneaking) {
                if (dolphin.value) {
                    player.motionY += 0.03999999910593033 // regular jump speed
                } else {
                    player.motionY = 0.1

                    if (player.ridingEntity != null && player.ridingEntity !is EntityBoat) {
                        player.ridingEntity?.motionY = 0.3
                    }
                }
            }
        }

        listener<AddCollisionBoxToListEvent> {
            if (dolphin.value || mc.player == null || mc.player.isSneaking || mc.player.fallDistance > 3) return@listener

            if (
                it.block is BlockLiquid
                && it.entity is EntityBoat
                && isAboveBlock(mc.player, it.pos)
                && (isDrivenByPlayer(it.entity) || it.entity === mc.player)
                && !isInWater(mc.player)
                && (EntityUtils.isAboveWater(mc.player, false) || EntityUtils.isAboveWater(mc.player.ridingEntity, false))
            ) {
                val axisAlignedBB = WATER_WALK_AA.offset(it.pos)
                if (it.entityBox.intersects(axisAlignedBB)) it.collidingBoxes.add(axisAlignedBB)
                it.cancel()
            }
        }

        listener<PacketEvent.Send> {
            if (dolphin.value) return@listener

            if (it.packet is CPacketPlayer
                && EntityUtils.isAboveWater(mc.player, true)
                && !isInWater(mc.player)
                && !isAboveLand(mc.player)
                && mc.player.ticksExisted % 2 == 0
            ) {
                it.packet.y += 0.02
            }
        }
    }

    private fun isAboveLand(entity: Entity): Boolean {
        val y = entity.posY - 0.01

        for (x in entity.posX.floorToInt() until entity.posX.ceilToInt()) {
            for (z in entity.posZ.floorToInt() until entity.posZ.ceilToInt()) {
                val pos = BlockPos(x, y.floorToInt(), z)
                if (mc.world.getBlockState(pos).isFullBlock) return true
            }
        }

        return false
    }

    private fun isAboveBlock(entity: Entity, pos: BlockPos): Boolean {
        return entity.posY >= pos.y
    }

    private fun isInWater(entity: Entity): Boolean {
        val y = (entity.posY + 0.01).floorToInt()

        for (x in entity.posX.floorToInt() until entity.posX.ceilToInt()) {
            for (z in entity.posZ.floorToInt() until entity.posZ.ceilToInt()) {
                val pos = BlockPos(x, y, z)
                if (mc.world.getBlockState(pos).block is BlockLiquid) return true
            }
        }

        return false
    }

    private fun isDrivenByPlayer(entity: Entity) = mc.player != null && entity == mc.player.ridingEntity
}